src/leo_s3_auth.erl

```diff
@@ -287,11 +315,83 @@ authenticate(Authorization, #sign_params{bucket = Bucket} = SignParams, IsCreate
                         <<"response-content-disposition">>,
                         <<"response-content-encoding">>]).
 
-%% @doc Get AWS signature version 2
--spec(get_signature(SecretAccessKey, SignParams) ->
+-spec(get_signature(SecretAccessKey, SignParams, SignV4Params) ->
              binary() when SecretAccessKey::binary(),
-                           SignParams::#sign_params{}).
-get_signature(SecretAccessKey, SignParams) ->
+                           SignParams::#sign_params{},
+                           SignV4Params::#sign_v4_params{}).
+get_signature(SecretAccessKey, SignParams, SignV4Params) ->
+    ?debug("get_signature/3", "Key: ~p, Sign: ~p, SignV4: ~p", [SecretAccessKey, SignParams, SignV4Params]),
+    case SignParams#sign_params.sign_ver of
+        v4 ->
+            get_signature_v4(SecretAccessKey, SignParams, SignV4Params);
+        _ ->
+            get_signature_v2(SecretAccessKey, SignParams)
+    end.
+
+%% @doc Get AWS signature version 4
+%% @private
+get_signature_v4(SecretAccessKey, SignParams, SignV4Params) ->
+    #sign_params{http_verb      = HTTPVerb,
+                 date           = Date,
+                 bucket         = Bucket,
+                 raw_uri        = URI,
+                 requested_uri  = RequestedURI,
+                 query_str      = QueryStr,
+                 req            = Req,
+                 amz_headers    = AmzHeaders
+                } = SignParams,
+    #sign_v4_params{credential      = Credential,
+                    signed_headers  = SignedHeaders
+                   } = SignV4Params,
+    URI_1       = auth_uri(Bucket, URI, RequestedURI),
+    Header_1    = auth_v4_headers(Req, SignedHeaders),
+    {Hash_1, _} = cowboy_req:header(<<"x-amz-content-sha256">>, Req),
+    Hash_2 = case Hash_1 of
+                 undefined ->
+                     %% Empty Payload
+                     <<"e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855">>;
+                 _ ->
+                     Hash_1
+             end,
+
+    QueryStr_1 = list_to_binary(auth_v4_qs(QueryStr)),
+
+    Request_1   = <<HTTPVerb/binary,        "\n",
+                    URI_1/binary,           "\n",
+                    QueryStr_1/binary,      "\n",
+                    Header_1/binary,        "\n",
+                    SignedHeaders/binary,   "\n",
+                    Hash_2/binary>>,
+
+    ?debug("get_signature_v4/3", "Request: ~p", [Request_1]),
+    RequestHash = crypto:hash(sha256, Request_1),
+    RequestHex  = leo_hex:binary_to_hex(RequestHash),
```

Comment:
Please implement leo_hex:binary_to_hex the return type binary version and replace with that to reduce unnecessary list_to_binary calls.
https://github.com/leo-project/leo_commons/blob/develop/src/leo_hex.erl